src/main/java/io/supertokens/useridmapping/UserIdMapping.java

```diff
@@ -17,20 +17,73 @@
+import io.supertokens.pluginInterface.session.SessionStorage;
 import io.supertokens.pluginInterface.useridmapping.UserIdMappingStorage;
 import io.supertokens.pluginInterface.useridmapping.exception.UnknownSuperTokensUserIdException;
 import io.supertokens.pluginInterface.useridmapping.exception.UserIdMappingAlreadyExistsException;
+import io.supertokens.pluginInterface.usermetadata.UserMetadataStorage;
+import io.supertokens.pluginInterface.userroles.UserRolesStorage;
 import io.supertokens.storageLayer.StorageLayer;
+import io.supertokens.userroles.UserRoles;
 
 import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.HashMap;
 
 public class UserIdMapping {
 
     public static void createUserIdMapping(Main main, String superTokensUserId, String externalUserId,
             String externalUserIdInfo)
             throws UnknownSuperTokensUserIdException, UserIdMappingAlreadyExistsException, StorageQueryException {
+        createUserIdMapping(main, superTokensUserId, externalUserId, externalUserIdInfo, false);
+    }
+
+    public static void createUserIdMapping(Main main, String superTokensUserId, String externalUserId,
+            String externalUserIdInfo, boolean force)
+            throws UnknownSuperTokensUserIdException, UserIdMappingAlreadyExistsException, StorageQueryException {
+        // if a userIdMapping is created with force, then we skip the following checks
+        if (!force) {
+            // check that none of the non-auth recipes are using the superTokensUserId
+            {
+
+                if (StorageLayer.getStorage(main).isUserIdBeingUsedInNonAuthRecipe(SessionStorage.class.getName(),
+                        superTokensUserId)) {
+                    throw new IllegalStateException("SuperTokens Id is already in use in Session recipe");
+                }
+            }
+
+            {
+                if (StorageLayer.getStorage(main).isUserIdBeingUsedInNonAuthRecipe(UserMetadataStorage.class.getName(),
+                        superTokensUserId)) {
+                    throw new IllegalStateException("SuperTokens Id is already in use in UserMetadata recipe");
+                }
+            }
+
+            {
+                if (StorageLayer.getStorage(main).isUserIdBeingUsedInNonAuthRecipe(UserRolesStorage.class.getName(),
+                        superTokensUserId)) {
+                    throw new IllegalStateException("SuperTokens Id is already in use in UserRoles recipe");
+                }
+            }
+            {
+                if (StorageLayer.getStorage(main).isUserIdBeingUsedInNonAuthRecipe(
+                        EmailVerificationStorage.class.getName(), superTokensUserId)) {
+                    throw new IllegalStateException("SuperTokens Id is already in use in EmailVerification recipe");
+                }
+            }
+
+            // Do not allow a mapping when the externalId is the superTokensUserId for another User
+            {
```

Comment:
add proper comment explaining this and also that the invalid state is possible in another way which we are consciously deciding not to worry about